src/web/mjs/connectors/mangaz.mjs

```diff
@@ -0,0 +1,90 @@
+import Connector from '../engine/Connector.mjs';
+import Manga from '../engine/Manga.mjs';
+
+export default class Mangaz extends Connector {
+
+    constructor() {
+        super();
+        super.id = 'mangaz';
+        super.label = 'Manga Library Z (マンガ図書館Z)';
+        this.tags = ['manga', 'japanese'];
+        this.url = 'https://www.mangaz.com';
+    }
+
+    async _getMangas() {
+        let mangaList = [];
+        for(let page = 0, run = true; run; page++) {
+            let mangas = await this._getMangasFromPage(page);
+            mangas.length > 0 ? mangaList.push(...mangas) : run = false;
+        }
+        return mangaList;
+    }
+
+    async _getMangasFromPage(page) {
+        const request = new Request(new URL('/title/addpage_renewal?query=&page='+page, this.url), {
+            method:'GET',
+            headers: {
+                'X-Requested-With': 'XMLHttpRequest'
+            }
+        });
+        const data = await this.fetchDOM(request, 'h4 > a');
+        return data.map(element => {
+            return {
+                id: this.getRootRelativeOrAbsoluteLink(element, request.url),
+                title: element.text.trim()
+            };
+        });
+    }
+
+    async _getMangaFromURI(uri) {
+        const request = new Request(uri, this.requestOptions);
+        const data = await this.fetchDOM(request, 'li.title');
+        const id = uri.pathname + uri.search;
+        const title = data[0].textContent.trim();
+        return new Manga(this, id, title);
+    }
+
+    async _getPages(chapter) {
+        const script = `
+            new Promise(async (resolve,reject) => {
+                let g = JCOMI.namespace("JCOMI.document")
+                let b = g.getDoc()
+                let img = g.getImages().map(ele => g.getLocationDir('enc') + ele.file + "?vw=" + encodeURIComponent(JCOMI.namespace("JCOMI.config").getVersion()))
+                resolve({img:img,b:b});
+            });
+        `;
+        const request = new Request(new URL(chapter.id, this.url), this.requestOptions);
+        const data = await Engine.Request.fetchUI(request, script)
+        return data.img.map(ele => this.createConnectorURI({
+            url:this.getAbsolutePath(ele, request.url),
+            key:data.b.Enc.key,
+            iv:data.b.Enc.iv
+        }));
+    }
+
+    async _handleConnectorURI(payload) {
+        let response = await fetch(payload.url)
+        let encrypted = await response.text();
+        let key = payload.key;
+        let iv = payload.iv;
+        let data = CryptoJS.AES.decrypt(encrypted, key, {iv:iv});
+        return {
+            mimeType: response.headers.get('content-type'),
+            data: data
```

Comment:
Isn't data still a base64 encoded string at this point?
Therefore it probably needs to be converted to `Uint8Array` => https://stackoverflow.com/a/41106346

Reply:
it doesn't output a string , i don't even know if its outputting the right thing

![](https://i.imgur.com/Vw4og4m.png)

Reply:
Since the decrypted data is a `WordArray` it needs to be converted to `Uint8Array`

Reply:
i have tried every combination of every different way used by anyone on google and the CryptoJS docs, i have no idea how to get anything without forge

Reply:
Try with this example (you can throw this directly in HakuNeko's console for testing)
```javascript
(async function handleConnectorURI(payload) {
    const response = await fetch(payload.url);
    const encrypted = CryptoJS.lib.WordArray.create(await response.arrayBuffer());
    const iv = CryptoJS.enc.Base64.parse(btoa(payload.iv));
    const key = CryptoJS.enc.Utf8.parse(payload.key);
    let decrypted = CryptoJS.AES.decrypt({ ciphertext: encrypted }, key, { iv: iv });
    decrypted = decrypted.toString(CryptoJS.enc.Utf8);
    decrypted = Uint8Array.from(atob(decrypted), char => char.charCodeAt(0));
    return {
        mimeType: response.headers.get('content-type'),
        data: decrypted
    };
})({
    url: 'https://cf.mangaz.com/Books/158/158634/xb_D86f9/1_Maccf.xb?vw=6.5.2',
    iv: atob('fy9utLFsZVoKgjxayZ1fzw=='),
    key: 'a76819c5740c6ce6'
});
```

Reply:
If you really want to use forge you need to download the [minified distribution](https://cdnjs.cloudflare.com/ajax/libs/forge/0.10.0/forge.min.js) to `./src/web/js` and include it in `./src/web/index.html`.

However, i don't see a point in using two different encryption libraries.

Furthermore i performed a benchmark with an encrypted 200 KB image where `forge` decryption time was ~`25 ms` and crypto-js was ~ `30 ms`, both are very fast and the difference is negligible compared to the time it takes to download the image data from the remote server.

Reply:
**Ronny Solution is working** 

![image](https://user-images.githubusercontent.com/31276236/98607059-f4e98a80-231a-11eb-983f-e46345e02c35.png)